File: php/src/DataStructures/SinglyLinkedList.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Singly linked list implementation.
 * Time: O(1) front ops, O(n) back ops and search
 * Space: O(n)
 *
 * @template T
 */
class SinglyLinkedList
{
    private ?SinglyNode $head = null;
    private ?SinglyNode $tail = null;
    private int $size = 0;

    public function pushFront(mixed $value): void
    {
        $node = new SinglyNode($value, $this->head);
        $this->head = $node;
        $this->tail ??= $node;
        $this->size++;
    }

    public function pushBack(mixed $value): void
    {
        $node = new SinglyNode($value);
        if ($this->tail !== null) {
            $this->tail->next = $node;
        } else {
            $this->head = $node;
        }
        $this->tail = $node;
        $this->size++;
    }

    public function popFront(): mixed
    {
        if ($this->head === null) {
            return null;
        }
        $value = $this->head->value;
        $this->head = $this->head->next;
        if ($this->head === null) {
            $this->tail = null;
        }
        $this->size--;
        return $value;
    }

    public function peekFront(): mixed
    {
        return $this->head?->value;
    }

    public function peekBack(): mixed
    {
        return $this->tail?->value;
    }

    public function contains(mixed $value): bool
    {
        $current = $this->head;
        while ($current !== null) {
            if ($current->value === $value) {
                return true;
            }
            $current = $current->next;
        }
        return false;
    }

    public function remove(mixed $value): bool
    {
        if ($this->head === null) {
            return false;
        }
        if ($this->head->value === $value) {
            $this->head = $this->head->next;
            if ($this->head === null) {
                $this->tail = null;
            }
            $this->size--;
            return true;
        }
        $current = $this->head;
        while ($current->next !== null) {
            if ($current->next->value === $value) {
                if ($current->next === $this->tail) {
                    $this->tail = $current;
                }
                $current->next = $current->next->next;
                $this->size--;
                return true;
            }
            $current = $current->next;
        }
        return false;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function size(): int
    {
        return $this->size;
    }

    public function toArray(): array
    {
        $result = [];
        $current = $this->head;
        while ($current !== null) {
            $result[] = $current->value;
            $current = $current->next;
        }
        return $result;
    }

    public function clear(): void
    {
        $this->head = null;
        $this->tail = null;
        $this->size = 0;
    }
}

/**
 * @internal
 */
class SinglyNode
{
    public function __construct(
        public mixed $value,
        public ?SinglyNode $next = null
    ) {}
}

File: csharp/src/DataStructures/SinglyLinkedList.cs
namespace DSA.DataStructures;

/// <summary>
/// Singly linked list implementation.
/// Time: O(1) front ops, O(n) back ops and search
/// Space: O(n)
/// </summary>
public class SinglyLinkedList<T>
{
    private class Node
    {
        public T Value { get; set; }
        public Node? Next { get; set; }

        public Node(T value)
        {
            Value = value;
        }
    }

    private Node? _head;
    private Node? _tail;
    private int _size;

    public int Count => _size;
    public bool IsEmpty => _size == 0;

    public void PushFront(T value)
    {
        var node = new Node(value) { Next = _head };
        _head = node;
        _tail ??= node;
        _size++;
    }

    public void PushBack(T value)
    {
        var node = new Node(value);
        if (_tail != null)
            _tail.Next = node;
        else
            _head = node;
        _tail = node;
        _size++;
    }

    public T? PopFront()
    {
        if (_head == null)
            return default;

        var value = _head.Value;
        _head = _head.Next;
        if (_head == null)
            _tail = null;
        _size--;
        return value;
    }

    public T? PeekFront() => _head == null ? default : _head.Value;
    public T? PeekBack() => _tail == null ? default : _tail.Value;

    public bool Contains(T value)
    {
        var current = _head;
        while (current != null)
        {
            if (EqualityComparer<T>.Default.Equals(current.Value, value))
                return true;
            current = current.Next;
        }
        return false;
    }

    public bool Remove(T value)
    {
        if (_head == null)
            return false;

        if (EqualityComparer<T>.Default.Equals(_head.Value, value))
        {
            _head = _head.Next;
            if (_head == null)
                _tail = null;
            _size--;
            return true;
        }

        var current = _head;
        while (current.Next != null)
        {
            if (EqualityComparer<T>.Default.Equals(current.Next.Value, value))
            {
                if (current.Next == _tail)
                    _tail = current;
                current.Next = current.Next.Next;
                _size--;
                return true;
            }
            current = current.Next;
        }
        return false;
    }

    public List<T> ToList()
    {
        var result = new List<T>();
        var current = _head;
        while (current != null)
        {
            result.Add(current.Value);
            current = current.Next;
        }
        return result;
    }

    public void Clear()
    {
        _head = null;
        _tail = null;
        _size = 0;
    }
}
